include/deal.II/lac/petsc_ts.h

```diff
@@ -0,0 +1,706 @@
+                                            (P ? P->petsc_matrix() : nullptr),
+                                        P ? P->petsc_matrix() : nullptr,
+                                        _dealii_ts_rhsjacobian_,
+                                        nullptr));
+            }
+
+          if (implicit_jacobian)
+            {
+              AssertTS(TSSetIJacobian(ts,
+                                      A ? A->petsc_matrix() :
+                                          (P ? P->petsc_matrix() : nullptr),
+                                      P ? P->petsc_matrix() : nullptr,
+                                      _dealii_ts_ijacobian_,
+                                      nullptr));
+            }
+        }
+
+
+      // In case solve_for_jacobian_system is provided, create a shell
+      // preconditioner wrapping the user call. The internal Krylov
+      // solver will apply the preconditioner only once. This choice
+      // can be overriden by command line and users can use any other
+      // Krylov method if their solve is not accurate enough.
+      PreconditionShell precond(
+        PetscObjectComm(reinterpret_cast<PetscObject>(ts)));
+      if (solve_for_jacobian_system)
+        {
+          precond.vmult = [&](VectorBase &      indst,
+                              const VectorBase &insrc) -> int {
+            VectorType       dst(static_cast<const Vec &>(indst));
+            const VectorType src(static_cast<const Vec &>(insrc));
+            return solve_for_jacobian_system(src, dst);
+          };
+
+          SNES snes;
+          KSP  ksp;
+          AssertTS(TSGetSNES(ts, &snes));
+          AssertTS(SNESGetKSP(snes, &ksp));
+          AssertTS(KSPSetType(ksp, KSPPREONLY));
+          AssertTS(KSPSetPC(ksp, precond.get_pc()));
+        }
+
+      // Attach user monitoring routine
+      if (monitor)
+        AssertTS(TSMonitorSet(ts, _dealii_ts_monitor_, nullptr, nullptr));
+
+      AssertTS(TSSetFromOptions(ts));
+
+      AssertTS(TSSolve(ts, nullptr));
+
+      PetscInt nt;
+      AssertTS(TSGetStepNumber(ts, &nt));
+      return nt;
+    }
+
+    /**
+     * Integrate the differential-algebraic equations.
+     * Here we also pass the matrices to handle Jacobians.
+     */
+    unsigned int
+    solve(VectorType &solution, PMatrixType &P)
+    {
+      reinit_matrices(P);
+      return solve(solution);
+    }
+
+    /**
+     * Integrate the differential-algebraic equations.
+     * Here we also pass the matrices to handle Jacobians.
+     */
+    unsigned int
+    solve(VectorType &solution, AMatrixType &A, PMatrixType &P)
+    {
+      reinit_matrices(A, P);
+      return solve(solution);
+    }
+
+    /**
+     * Compute implicit residual $F(t, y, \dot y)$.
+     */
+    std::function<int(const double      t,
+                      const VectorType &y,
+                      const VectorType &y_dot,
+                      VectorType &      res)>
+      implicit_function;
```

Comment:
Could we give these function objects a name that includes a verb? They represent actions, after all. So here maybe `evaluate_implicit_residual()`?

Reply:
Ah, ok. Then that train has left the station already :-(